src_c/camera_windows.c

```diff
@@ -0,0 +1,859 @@
+    HANDLEHR(hr);
+
+    *ppAttr = res;
+    return 1;
+
+    cleanup:
+        free(res);
+        return 0;
+}
+
+WCHAR **
+windows_list_cameras(int *num_devices)
+{
+    WCHAR** devices = NULL;
+    IMFAttributes *pAttributes = NULL;
+    IMFActivate **ppDevices = NULL;
+    UINT32 count = 0;
+    HRESULT hr;
+
+    hr = MFCreateAttributes(&pAttributes, 1);
+    HANDLEHR(hr);
+
+    hr = pAttributes->lpVtbl->SetGUID(pAttributes,
+                                      &MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+                                      &DEVSOURCE_VIDCAP_GUID);
+
+    HANDLEHR(hr);
+
+    hr = MFEnumDeviceSources(pAttributes, &ppDevices, &count);
+    HANDLEHR(hr);
+
+    /* freed by caller of this function */
+    devices = (WCHAR **)malloc(sizeof(WCHAR *) * count);
+    if (!devices) {
+        hr = E_OUTOFMEMORY;
+        HANDLEHR(hr);
+    }
+
+    for (int i=0; i<(int)count; i++) {
+        if(!_get_name_from_activate(ppDevices[i], &devices[i])) {
+            hr = HR_UPSTREAM_FAILURE;
+            goto cleanup;
+        }
+    }
+
+    *num_devices = count;
+    goto cleanup;
+
+    cleanup:
+        RELEASE(pAttributes);
+        for (int i=0; i<(int)count; i++) {
+            RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+        if FAILED(hr) {
+            for (int i=0; i<(int)count; i++) {
+                free(devices[i]);
+            }
+            free(devices);
+            return NULL;
+        }
+        else {
+            return devices;
+        }
+}
+
+IMFActivate *
+windows_device_from_name(WCHAR* device_name)
+{
+    IMFAttributes *pAttributes = NULL;
+    IMFActivate **ppDevices = NULL;
+    WCHAR* _device_name = NULL;
+    UINT32 count = 0;
+    HRESULT hr;
+
+    hr = MFCreateAttributes(&pAttributes, 1);
+    HANDLEHR(hr);
+
+    hr = pAttributes->lpVtbl->SetGUID(pAttributes,
+                                      &MF_DEVSOURCE_ATTRIBUTE_SOURCE_TYPE,
+                                      &DEVSOURCE_VIDCAP_GUID);
+    HANDLEHR(hr);
+
+    hr = MFEnumDeviceSources(pAttributes, &ppDevices, &count);
+    HANDLEHR(hr);
+
+    for (int i=0; i<(int)count; i++) {
+        if (!_get_name_from_activate(ppDevices[i], &_device_name)) {
+            hr = HR_UPSTREAM_FAILURE;
+            goto cleanup;
+        }
+        if (!wcscmp(_device_name, device_name)) {
+            free(_device_name);
+            for (int j=0; j<(int)count; j++) {
+                if (i != j) {
+                    RELEASE(ppDevices[j]);
+                }
+            }
+            CoTaskMemFree(ppDevices);
+            return ppDevices[i];
+        }
+        free(_device_name);
+    }
+
+    goto cleanup;
+
+    cleanup:
+        RELEASE(pAttributes);
+        for (int i=0; i<(int)count; i++) {
+            RELEASE(ppDevices[i]);
+        }
+        CoTaskMemFree(ppDevices);
+        return NULL;
+}
+
+int
+_select_source_type(pgCameraObject* self, IMFMediaType **mp)
```

Comment:
Could you please describe to me what this function does? It's a bit unclear to me.

Reply:
I will definitely add some more comments for some of these functions.

This enumerates the "native modes" of the selected camera, which include things like resolution, format, and FPS. It then tries to select the option that is closest to the requested camera size, and has an uncompressed format. (MJPEG, MP4 would be compressed - YUV/RGB would be uncompressed). See: https://docs.microsoft.com/en-us/windows/win32/medfound/video-subtype-guids

Specifically, it goes for not just any uncompressed format, but ones that are supported input types for the video processor MFT, which does conversion from "random uncompressed" to RGB32 / YUV for reading the results.

If it can't find a single supported format in any mode, it explicitly errors. But it always should, because in Windows 10 Microsoft some additional decoding behind the scenes for efficiency when multiple programs read the camera.